### PR TITLE
String storage fix

### PR DESCRIPTION
**Pull request by nwatson22**
Adds structuring assumptions to strings in contract storage when running proofs in summary mode, allowing these proofs to finish.
Specifically, assumes for all storage slots that hold a `string`, that the least significant bit is set, which signals the length is <32 bytes, and that the length, which is stored in this case in the next 7 bits, actually <32, and is positive.

### CODE REVIEW

File: src/kontrol/prove.py [hunk 1]
```diff
@@ -869,9 +869,10 @@ def _init_cterm(
             # Symbolic account for the contract being executed
             accounts.append(Foundry.symbolic_account(Foundry.symbolic_contract_prefix(), contract_code))
         else:
+            storage_map = KVariable(Foundry.symbolic_contract_prefix() + '_STORAGE', sort=KSort('Map'))
```
Comment (palinatolmach) [inline src/kontrol/prove.py:872]: Is this variable declaration needed here and as a separate argument to `_create_cse_accounts`? It gets initialized in the same manner within `_create_cse_accounts`: https://github.com/runtimeverification/kontrol/blob/fe577e524655da12a6313c3baa4ed671890b23ed/src/kontrol/prove.py#L977
On the first (non-recursive) invocation of `_create_cse_accounts`, `contract_name` is `Foundry.symbolic_contract_prefix()`, so it gets initialized with the exact same values, I think.

File: src/tests/integration/test_kontrol_cse.py [hunk 2]
```diff
@@ -48,6 +48,10 @@ def test_foundry_dependency_automated(
     if bug_report is not None:
         server._populate_bug_report(bug_report)
 
+    run_constructor = True
+    if test_id in ['CallableStorageTest.test_str()', 'CallableStorageContract.str()']:
```
Comment (palinatolmach) [inline src/tests/integration/test_kontrol_cse.py:52]: Would `CallableStorageContract.init` fail if we run it with `--run-constructor`? Would it be sufficient to handle the engagement contracts @PetarMax?

File: src/kontrol/prove.py [hunk 3]
```diff
@@ -972,6 +978,63 @@ def _create_cse_accounts(
     new_accounts.append(Foundry.symbolic_account(contract_name, contract_code, storage_map))
 
     for field in storage_fields:
+        if field.data_type == 'string':
+            lookup = KApply('lookup', [storage_map, intToken(field.slot)])
+            length_byte_lt32 = KApply(
+                '_<Int_',
```
[the remark this answers is not in the document]
Comment (nwatson22) [reply src/kontrol/prove.py]: Fixed.

File: src/kontrol/prove.py [hunk 4]
```diff
@@ -972,6 +978,63 @@ def _create_cse_accounts(
...
+                                    intToken(1),
+                                ],
+                            ),
+                        ],
+                    ),
+                    intToken(0),
+                ],
+            )
+            lowest_bit_not_set = KApply(
+                '_==Int_',
```
[the remark this answers is not in the document]
Comment (nwatson22) [reply src/kontrol/prove.py]: Fixed.

File: src/kontrol/prove.py (hunk 1 above)
Comment (nwatson22) [reply src/kontrol/prove.py]: Fixed.

File: src/tests/integration/test_kontrol_cse.py (hunk 2 above)
Comment (PetarMax) [reply src/tests/integration/test_kontrol_cse.py:52]: We don't want to handle this. We need the constraints in constructors as well, either as part of this PR or the next one.

File: src/tests/integration/test-data/foundry/test/CallableStorageTest.t.sol [hunk 5]
```diff
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity =0.8.13;
+
+import "forge-std/Test.sol";
+import "kontrol-cheatcodes/KontrolCheats.sol";
+
+contract CallableStorageContract {
+    uint public num;
+    string public str;
+    uint immutable imm;
```
Comment (PetarMax) [inline src/tests/integration/test-data/foundry/test/CallableStorageTest.t.sol:10]: Is this needed?

File: src/tests/integration/test-data/foundry/test/CallableStorageTest.t.sol [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+    constructor(string memory a) payable {
+        str = a;
+        imm = 2;
+    }
+
+    function doNothing() public {
+        return;
+    }
```
Comment (PetarMax) [inline src/tests/integration/test-data/foundry/test/CallableStorageTest.t.sol:19]: This, too...

File: src/tests/integration/test-data/foundry/test/CallableStorageTest.t.sol (hunk 5 above)
Comment (nwatson22) [reply src/tests/integration/test-data/foundry/test/CallableStorageTest.t.sol]: Removed.

File: src/tests/integration/test-data/foundry/test/CallableStorageTest.t.sol (hunk 6 above)
Comment (nwatson22) [reply src/tests/integration/test-data/foundry/test/CallableStorageTest.t.sol]: Removed.

**Comment by nwatson22** [comment]
> Let's remove the immutable and the `doNothing` function. Putting the string conditions in for when a constructor is run can be done as part of this PR as as part of next PR.

@PetarMax It's removed. I will save the string conditions for constructors for another PR. Is there something with constructors specifically with string arguments, because I thought it was the same issue for constructors and functions both (that it creates a copy loop).

**Comment by PetarMax** [comment]
No, there isn't, but I am guessing the constraints aren't put in at the moment at constructor-time?

**Review by PetarMax — APPROVED** [review]

**Pull request merged**